src/ZODB/DB.py

```diff
@@ -1081,15 +1106,23 @@ def tpc_vote(self, transaction):
         self._storage.tpc_vote(transaction)
 
     def tpc_finish(self, transaction):
-        transaction = transaction.data(self)
-        self._storage.tpc_finish(transaction)
+        with contextlib.closing(self):
```

Comment:
The only case where `contextlib.closing` is nice is when opening several objects in a row, which means one have a single `with` statement and single indentation, rather than nested try...finally.

Here, it's abuse of sugar. The object is already open so you don't even save the line for opening. The result is slower code for questionable benefit about readability.

On the other side, the next 2 lines can be merged.

Reply:
```python
with closing(thing):
    do_stuff()
```

is still two lines shorter than 

```python
try:
    do_stuff()
finally:
    thing.close()
```

It's arguably more readable, but that's clearly a matter of opinion. `contextlib` is somewhat slower, but that's unlikely to matter here. 

There are a couple semantic differences, too, the most significant of which is that [a `with` block is safer from asynchronous exceptions](https://vorpus.org/blog/control-c-handling-in-python-and-trio/#id4) such as `KeyboardInterrupt` than a `try/finally` pair. That's *also* probably unlikely to matter much here, though.

I don't much care so I went ahead and changed it.

Reply:
Thanks.

> There are a couple semantic differences, too, the most significant of which is that [a `with` block is safer from asynchronous exceptions](https://vorpus.org/blog/control-c-handling-in-python-and-trio/#id4) such as `KeyboardInterrupt` than a `try/finally` pair. That's _also_ probably unlikely to matter much here, though.

The link is interesting. I already knew the race about try...except, but not that `with` can be safer. If I read correctly, it's only safer if the work to undo is done in `__enter__` (not the case of `closing`), and if `__enter__` is written in C (again not the case of `closing`).

try...finally is not less safe here.